src/rust/core/proxy/src/process.rs

```diff
@@ -0,0 +1,200 @@
+            .collect();
+
+        let frontends: Vec<FrontendWorker<RequestParser, Request, Response>> = frontend_builders
+            .drain(..)
+            .map(|v| {
+                v.build(
+                    signal_queue_rx.remove(0),
+                    queues_worker_session.remove(0),
+                    queues_frontend_data.remove(0),
+                )
+            })
+            .collect();
+        let listener = listener_builder.build(queues_listener_session.remove(0));
+
+        let admin = admin_builder.build(signal_queue_tx.remove(0), signal_rx);
+
+        Ok(Self {
+            admin,
+            listener,
+            frontends,
+            backends,
+            signal_tx,
+        })
+    }
+
+    #[allow(clippy::vec_init_then_push)]
+    pub fn spawn(mut self) -> Process {
+        let admin = std::thread::Builder::new()
+            .name("pelikan_admin".to_string())
+            .spawn(move || self.admin.run())
+            .unwrap();
+
+        let listener = std::thread::Builder::new()
+            .name("pelikan_listener".to_string())
+            .spawn(move || self.listener.run())
+            .unwrap();
+
+        let mut frontend = Vec::new();
+        for (id, fe) in self.frontends.drain(..).enumerate() {
+            frontend.push(
+                std::thread::Builder::new()
+                    .name(format!("pelikan_fe_{}", id))
+                    .spawn(move || fe.run())
+                    .unwrap(),
+            )
+        }
+
+        let mut backend = Vec::new();
+        for (id, be) in self.backends.drain(..).enumerate() {
+            backend.push(
+                std::thread::Builder::new()
+                    .name(format!("pelikan_be_{}", id))
+                    .spawn(move || be.run())
+                    .unwrap(),
+            )
+        }
+
+        Process {
+            admin,
+            listener,
+            frontend,
+            backend,
+            signal_tx: self.signal_tx,
+        }
+    }
+}
+
+pub struct Process {
+    admin: JoinHandle<()>,
+    listener: JoinHandle<()>,
+    backend: Vec<JoinHandle<()>>,
+    frontend: Vec<JoinHandle<()>>,
+    signal_tx: Sender<Signal>,
+}
+
+impl Process {
+    /// Attempts to gracefully shutdown the `Process` by sending a shutdown to
+    /// each thread and then waiting to join those threads.
+    ///
+    /// Will terminate ungracefully if it encounters an error in sending a
+    /// shutdown to any of the threads.
+    ///
+    /// This function will block until all threads have terminated.
+    pub fn shutdown(self) {
+        // this sends a shutdown to the admin thread, which will broadcast the
+        // signal to all sibling threads in the process
+        if self.signal_tx.try_send(Signal::Shutdown).is_err() {
+            fatal!("error sending shutdown signal to thread");
+        }
+
+        // wait and join all threads
+        self.wait()
```

Comment:
How do FE/BE workers handle shutdown signal currently? I imagine to properly flush all requests that are already in-flight will probably require some choreography like half-close FE pipe->half-close BE pipe->fully-close BE pipe->fully-close-FE pipe?

Reply:
Currently, no attempt is made to properly handle this. I think this is worthy of being moved into an issue and focus a PR on just this aspect. It definitely requires some careful thought and definitions around what proper handling should look like. I suspect it'll always be best-effort to handle requests in-flight, as we'll likely need to set some sort of timeout for handling those gracefully.

Reply:
#422